nifi-nar-bundles/nifi-cassandra-bundle/nifi-cassandra-processors/src/test/java/org/apache/nifi/processors/cassandra/PutCassandraQLTest.java

```diff
@@ -0,0 +1,219 @@
+
+    @Before
+    public void setUp() throws Exception {
+        processor = new MockPutCassandraQL();
+        testRunner = TestRunners.newTestRunner(processor);
+    }
+
+    @Test
+    public void testProcessorConfigValidity() {
+        testRunner.setProperty(AbstractCassandraProcessor.CONTACT_POINTS, "localhost:9042");
+        testRunner.assertValid();
+        testRunner.setProperty(AbstractCassandraProcessor.PASSWORD, "password");
+        testRunner.assertNotValid();
+        testRunner.setProperty(AbstractCassandraProcessor.USERNAME, "username");
+        testRunner.setProperty(AbstractCassandraProcessor.CONSISTENCY_LEVEL, "ONE");
+        testRunner.assertValid();
+    }
+
+    @Test
+    public void testProcessorHappyPath() {
+        setUpStandardTestConfig();
+
+        testRunner.run(1, true, true);
+        testRunner.assertAllFlowFilesTransferred(PutCassandraQL.REL_SUCCESS, 1);
+        testRunner.clearTransferState();
+    }
+
+    @Test
+    public void testProcessorInvalidQueryException() {
+        setUpStandardTestConfig();
+
+        // Test exceptions
+        processor.setExceptionToThrow(
+                new InvalidQueryException(new InetSocketAddress("localhost", 9042), "invalid query"));
+        testRunner.enqueue("UPDATE users SET cities = [ 'New York', 'Los Angeles' ] WHERE user_id = 'coast2coast';");
+        testRunner.run(1, true, true);
+        testRunner.assertAllFlowFilesTransferred(PutCassandraQL.REL_FAILURE, 1);
+        testRunner.clearTransferState();
+    }
+
+    @Test
+    public void testProcessorUnavailableException() {
+        setUpStandardTestConfig();
+
+        processor.setExceptionToThrow(
+                new UnavailableException(new InetSocketAddress("localhost", 9042), ConsistencyLevel.ALL, 5, 2));
+        testRunner.enqueue("UPDATE users SET cities = [ 'New York', 'Los Angeles' ] WHERE user_id = 'coast2coast';");
+        testRunner.run(1, true, true);
+        testRunner.assertAllFlowFilesTransferred(PutCassandraQL.REL_RETRY, 1);
+    }
+
+    @Test
+    public void testProcessorNoHostAvailableException() {
+        setUpStandardTestConfig();
+
+        processor.setExceptionToThrow(new NoHostAvailableException(new HashMap<InetSocketAddress, Throwable>()));
+        testRunner.enqueue("UPDATE users SET cities = [ 'New York', 'Los Angeles' ] WHERE user_id = 'coast2coast';");
+        testRunner.run(1, true, true);
+        testRunner.assertAllFlowFilesTransferred(PutCassandraQL.REL_RETRY, 1);
+    }
+
+    @Test
+    public void testProcessorProcessException() {
+        setUpStandardTestConfig();
+
+        processor.setExceptionToThrow(new ProcessException());
+        testRunner.enqueue("UPDATE users SET cities = [ 'New York', 'Los Angeles' ] WHERE user_id = 'coast2coast';");
+        testRunner.run(1, true, true);
+        testRunner.assertAllFlowFilesTransferred(PutCassandraQL.REL_FAILURE, 1);
+    }
+
+    private void setUpStandardTestConfig() {
```

Comment:
In many tests "setUpStandardTestConfig" is used (enqueuing a flowfile), then another flowfile is enqueued and then testRunner is only run once. The second FlowFile is not needed.